node/circuits.js

```diff
@@ -0,0 +1,213 @@
+    var self = this;
+    var circuits = self.circuitsByCallerName['$' + callerName];
+    if (!circuits) {
+        circuits = new EndpointCircuits(self.root);
+        self.circuitsByCallerName['$' + callerName] = circuits;
+    }
+    return circuits.getCircuit(callerName, serviceName, endpointName);
+};
+
+function Circuits(options) {
+    var self = this;
+    self.circuitsByServiceName = {};
+    self.stateOptions = {
+        nextHandler: alwaysShouldRequestHandler,
+        timers: options.timers,
+        random: options.random,
+        period: options.period,
+        maxErrorRate: options.maxErrorRate,
+        minRequests: options.minRequests,
+        probation: options.probation
+    };
+    self.shouldRequestOptions = {};
+}
+
+Circuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    var circuits = self.circuitsByServiceName['$' + serviceName];
+    if (!circuits) {
+        circuits = new EndpointCircuits(self);
+        self.circuitsByServiceName['$' + serviceName] = circuits;
+    }
+    return circuits.getCircuit(callerName, serviceName, endpointName);
+};
+
+Circuits.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    // Default the caller name.
+    // All callers that fail to specifiy a cn share a circuit for each sn:en
+    // and fail together.
+    var callerName = req.headers.cn || 'yunocn';
+    var serviceName = req.serviceName;
+    if (!serviceName) {
+        return buildRes().sendError('BadRequest', 'All requests must have a service name');
+    }
+    return req.withArg1(function withArg1(endpointName) {
+        var circuit = self.getCircuit(callerName, serviceName, endpointName);
+        return circuit.handleRequest(req, buildRes, nextHandler);
+    });
+};
+
+// Called upon membership change to collect services that the corresponding
+// exit node is no longer responsible for.
+Circuits.prototype.updateServices = function updateServices(managesService) {
+    var self = this;
+    var serviceNames = Object.keys(self.circuitsByServiceName);
+    for (var index = 0; index < serviceNames.length; index++) {
+        var serviceName = serviceNames[index];
+        if (!managesService(serviceName)) {
+            delete self.circuitsByServiceName[serviceName];
+        }
+    }
+};
+
+function Circuit() {
+    var self = this;
+    self.circuitName = null;
+    self.callerName = null;
+    self.serviceName = null;
+    self.endpointName = null;
+    self.shouldRequestOptions = null;
+    self.stateOptions = null;
+    StateMachine.call(self);
+    EventEmitter.call(self);
+    self.stateChangedEvent = self.defineEvent('stateChanged');
+}
+
+inherits(Circuit, EventEmitter);
+
+Circuit.prototype.setState = StateMachine.prototype.setState;
+
+Circuit.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
```

Comment:
I can’t envision that working. We need a hook somewhere that enables the circuit breaker to observe or intercept the request and response. This will still be true with the pass-through frame relay when we have it. All that I suspect will change is that the content of arg2 and arg3 will not be retained on the request and response, which this adapter does not need anyway.